src/include/storage/index/index_builder.h

```diff
@@ -67,7 +72,85 @@ class IndexBuilder {
     return *this;
   }
 
+ /**
+  * @param sql_table, transaction context
+  * @return the builder object
+  */
+  IndexBuilder &SetSqlTableAndTransactionContext(common::ManagedPointer<storage::SqlTable> sql_table,
+                                                 const common::ManagedPointer<transaction::TransactionContext> txn) {
+    TERRIER_ASSERT((sql_table == nullptr && txn == nullptr) || (sql_table != nullptr && txn != nullptr),
+                   "sql_table / txn is null and txn / sql_table is not.");
+    sql_table_ = sql_table;
+    txn_ = txn;
+    return *this;
+  }
+ /**
+  *
+  * @param newly created index
+  * @return index with all the keys inserted
+  */
+  void BulkInsert(Index *index) const {
+    const auto index_pr_initializer = index->GetProjectedRowInitializer();
+
+    const uint32_t index_pr_size = index_pr_initializer.ProjectedRowSize();
+    byte *index_pr_buffer = common::AllocationUtil::AllocateAligned(index_pr_size);
+
+    ProjectedRow *index_pr = index_pr_initializer.InitializeRow(index_pr_buffer);
+
+    const auto &indexed_attributes = key_schema_.GetIndexedColOids();
+    const auto table_pr_initializer = sql_table_->InitializerForProjectedRow(indexed_attributes);
+
+    const uint32_t table_pr_size = table_pr_initializer.ProjectedRowSize();
+    byte *table_pr_buffer = common::AllocationUtil::AllocateAligned(table_pr_size);
+
+    ProjectedRow *table_pr = table_pr_initializer.InitializeRow(table_pr_buffer);
+
+    auto pr_map = sql_table_->ProjectionMapForOids(indexed_attributes);
+
+    for (auto it = sql_table_->begin(); it != sql_table_->end(); ++it) {
+      const TupleSlot slot = *it;
+
+      sql_table_->TraverseVersionChain(slot, table_pr, [&, this](auto type){
+        // Insert into the index
+        auto num_index_cols = key_schema_.GetColumns().size();
+        TERRIER_ASSERT(num_index_cols == indexed_attributes.size(), "Only support index keys that are a single column oid");
+        for (uint32_t col_idx = 0; col_idx < num_index_cols; col_idx++) {
+          const auto &col = key_schema_.GetColumn(col_idx);
+          auto index_col_oid = col.Oid();
+          const catalog::col_oid_t &table_col_oid = indexed_attributes[col_idx];
+          if (table_pr->IsNull(pr_map[table_col_oid])) {
+            index_pr->SetNull(index->GetKeyOidToOffsetMap().at(index_col_oid));
+          } else {
+            auto size = AttrSizeBytes(col.AttrSize());
+            std::memcpy(index_pr->AccessForceNotNull(index->GetKeyOidToOffsetMap().at(index_col_oid)),
+                        table_pr->AccessWithNullCheck(pr_map[table_col_oid]), size);
+          }
+        }
+
+        switch (type) {
+          case DataTable::VersionChainType::VISIBLE:
+            printf("Inserting into index");
+            index->Insert(txn_, *index_pr, slot);
+            break;
+          case DataTable::VersionChainType::INVISIBLE:
+          case DataTable::VersionChainType::PRE_UPDATE:
```

Comment:
what's the point of inserting then deleting?

Reply:
That's how updates are handled on an index - you insert the new value and then delete the old value

Reply:
Specifically, this registers the old value in the index and then registers a delete, which what the index does is defer this delete until later. This means currently running transactions timestamped before this update will see it in the index, but it will be removed once current transactions are finished.